File: src/web/Controllers/BrokerageController.cs
using System.IO;
using System.Text;
using System.Threading.Tasks;
using core.fs.Brokerage;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using web.Utils;

namespace web.Controllers
{
    [ApiController]
    [Route("api/[controller]")]
    [Authorize]
    public class BrokerageController(BrokerageHandler brokerageHandler) : ControllerBase
    {
        [HttpPost("buy")]
        public Task<ActionResult> Buy([FromBody] BuyOrSellData data) =>
            this.OkOrError(
                brokerageHandler.Handle(
                    BrokerageTransaction.NewBuy(data, User.Identifier())
                )
            );
        
        [HttpPost("buytocover")]
        public Task<ActionResult> BuyToCover([FromBody] BuyOrSellData data) =>
            this.OkOrError(
                brokerageHandler.Handle(
                    BrokerageTransaction.NewBuyToCover(data, User.Identifier())
                )
            );
        
        [HttpPost("sellshort")]
        public Task<ActionResult> SellShort([FromBody] BuyOrSellData data) =>
            this.OkOrError(
                brokerageHandler.Handle(
                    BrokerageTransaction.NewSellShort(data,User.Identifier())
                )
            );

        [HttpPost("sell")]
        public Task<ActionResult> Sell([FromBody] BuyOrSellData data) =>
            this.OkOrError(
                brokerageHandler.Handle(BrokerageTransaction.NewSell(data, User.Identifier()))
            );

        [HttpPost("optionsorder")]
        public async Task<ActionResult> OptionsOrder()
        {
            var reader = new StreamReader(Request.Body, Encoding.UTF8);

            var json = await reader.ReadToEndAsync();

            var result = brokerageHandler.Handle(new OptionOrderCommand(User.Identifier(), json));
            
            return await this.OkOrError(result);
        }

        [HttpDelete("orders/{orderId}")]
        public Task<ActionResult> Delete([FromRoute] string orderId) =>
            this.OkOrError(
                brokerageHandler.Handle(new CancelOrder(User.Identifier(), orderId)));

        [HttpGet("account")]
        public Task<ActionResult> GetAccount() =>
            this.OkOrError(
                brokerageHandler.Handle(new QueryAccount(User.Identifier())));
        
        [HttpGet("transactions")]
        public Task<ActionResult> GetTransactions() =>
            this.OkOrError(
                brokerageHandler.Handle(new QueryTransactions(User.Identifier())));
    }
}

File: src/web.fs/Controllers/BrokerageController.fs
namespace web.Controllers

open System.IO
open System.Text
open System.Threading.Tasks
open Microsoft.AspNetCore.Mvc
open Microsoft.AspNetCore.Authorization
open core.fs.Brokerage
open web.Utils

[<ApiController>]
[<Route("api/[controller]")>]
[<Authorize>]
type BrokerageController(brokerageHandler: BrokerageHandler) =
    inherit ControllerBase()

    [<HttpPost("buy")>]
    member this.Buy([<FromBody>] data: BuyOrSellData) =
        this.OkOrError(
            brokerageHandler.Handle(
                BrokerageTransaction.Buy(data, this.User.Identifier())
            )
        )

    [<HttpPost("buytocover")>]
    member this.BuyToCover([<FromBody>] data: BuyOrSellData) =
        this.OkOrError(
            brokerageHandler.Handle(
                BrokerageTransaction.BuyToCover(data, this.User.Identifier())
            )
        )

    [<HttpPost("sellshort")>]
    member this.SellShort([<FromBody>] data: BuyOrSellData) =
        this.OkOrError(
            brokerageHandler.Handle(
                BrokerageTransaction.SellShort(data, this.User.Identifier())
            )
        )

    [<HttpPost("sell")>]
    member this.Sell([<FromBody>] data: BuyOrSellData) =
        this.OkOrError(
            brokerageHandler.Handle(
                BrokerageTransaction.Sell(data, this.User.Identifier())
            )
        )

    [<HttpPost("optionsorder")>]
    member this.OptionsOrder() = task {
        use reader = new StreamReader(this.Request.Body, Encoding.UTF8)
        let! json = reader.ReadToEndAsync()
        let result = brokerageHandler.Handle({OptionOrderCommand.UserId = this.User.Identifier(); Payload = json})
        return! this.OkOrError(result)
    }

    [<HttpDelete("orders/{orderId}")>]
    member this.Delete([<FromRoute>] orderId: string) =
        this.OkOrError(
            brokerageHandler.Handle({CancelOrder.UserId = this.User.Identifier(); OrderId = orderId})
        )

    [<HttpGet("account")>]
    member this.GetAccount() =
        this.OkOrError(
            brokerageHandler.Handle({QueryAccount.UserId = this.User.Identifier()})
        )

    [<HttpGet("transactions")>]
    member this.GetTransactions() =
        this.OkOrError(
            brokerageHandler.Handle({QueryTransactions.UserId = this.User.Identifier()})
        )
